marxs/source/source.py

```diff
@@ -4,22 +4,102 @@
 
 from ..base import SimulationSequenceElement
 from ..optics.polarization import polarization_vectors
+from ..math.random import RandomArbitraryPdf
+
+def random_polarization(n, dir_array):
+    # randomly choose polarization (temporary)
+    angles = np.random.uniform(0, 2 * np.pi, n)
+    return polarization_vectors(dir_array, angles)
+
+
+class timedependentspectrum(object):
+    '''
+
+    Example
+    ------
+    >>> mysource = Source(energy=timedependedspectrum(2.3, 3.4))
+    '''
+    def __init__(self, en1, en2):
+        self.en1 = en1
+        self.en2 = en2
+
+    def __call__(self, t):
+        energies = np.zeros_like(t)
+        energies[t < 1] = self.en1
+        energies[t >=1] = self.en2
+        return energies
 
-'''TO-DO: make proper object hirachy, always call super on init. Pass args through, e.g. for name.'''
 
 class Source(SimulationSequenceElement):
     '''Make this ABC once I have worked out the interface'''
+    def __init__(self, **kwargs):
+        self.energy = kwargs.pop('energy')
+        self.flux = kwargs.pop('flux')
+        self.polarization = kwargs.pop('polarization', None)
+
+        super(Source, self).__init__(**kwargs)
+
+    def generate_times(self, exposuretime):
+        if callable(self.flux):
+            return self.rate(exposuretime)
+        elif np.isscalar(self.flux):
+            return np.arange(0, exposuretime, 1./self.flux)
+        else:
+            raise ValueError('Rate must be a number of a callable.')
```

Comment:
I think this should say `'Flux must be...'` instead of `Rate` to be clear. (And `of` should be `or`)